### DISCUSSION TITLE
Pattern Matching Functions in C#

### DISCUSSION THREAD

**Discussion by DavidArno** [dotnet/csharplang#6916] [Language Ideas]
In F#, a function that just does a pattern match:
```f#
let f aValue =
    match aValue with
    | _ -> "something"
```
can be expressed using "pattern matching function" notation:
```f#
let f =
    function
    | _ -> "something"
```

It would be nice to have a similar short-hand in C# so that:
```cs
enum Quadrant { TopLeft, TopRight, BottomLeft, BottomRight }

Quadrant CalcQuadrant(double x, double y) =>
    (x, y) switch {
        (< 0, >= 0) => Quadrant.TopLeft,
        (>= 0, >= 0) => Quadrant.TopRight,
        (< 0, < 0) => Quadrant.BottomLeft,
        _ => Quadrant.BottomRight
    };
```
could be expressed as:
```cs
Quadrant CalcQuadrant(double x, double y) => {
    (< 0, >= 0) => Quadrant.TopLeft,
    (>= 0, >= 0) => Quadrant.TopRight,
    (< 0, < 0) => Quadrant.BottomLeft,
    _ => Quadrant.BottomRight
};
```
As with the F# version, it just saves the `match x with` part (`(x, y) switch` in the case of this C# example). A small saving, but for those that write a lot of methods that only perform a pattern match, a useful one.

**Comment by HaloFour** [comment]
I'm not a fan of using the lambda operator here.  It wouldn't be ambiguous with methods, but it could be at least confusing with lambdas where `=>` followed by a block is legal and normal.  I think I'd prefer a syntax that would explicitly include the keyword `switch` somewhere.  Although at that point you're not really far from the current syntax to achieve this functionality.
